recommender/services/health_check.py

```diff
@@ -0,0 +1,157 @@
+
+import stomp
+from flask import current_app
+from mongoengine import Document
+
+from recommender.engines.base.base_inference_component import (
+    MLEngineInferenceComponent,
+    BaseInferenceComponent,
+)
+from recommender.engines.engines import ENGINES
+from recommender.errors import NoSavedMLComponentError
+from recommender.models import Service, User
+from recommender.tasks import ping
+
+# TODO: adjust if needed
+HEALTHY_USERS_THRESHOLD = 10
+# TODO: adjust if needed
+HEALTHY_SERVICES_THRESHOLD = 10
+K = 3
+
+
+class HealthStatus:
+    """
+    Class representing a health status of a component.
+    Components form a tree-like structure.
+    Can be a leaf (specifies status explicitly) or
+        a node (status is healthy if all the children are healthy)
+    """
+
+    name: str
+    status: bool
+    error: Optional[str]
+    components: List[HealthStatus]
+
+    def __init__(
+        self,
+        name: str,
+        status_or_components: Union[bool, List[HealthStatus]],
+        error: Optional[str] = None,
+    ):
+        self.name = name
+        self.error = error
+
+        if isinstance(status_or_components, bool):
+            self.status = status_or_components
+            self.components = []
+        else:
+            self.status = all(component.status for component in status_or_components)
+            self.components = status_or_components
+
+    def to_dict(self) -> Dict:
+        """Report the health status as dict."""
+        as_dict = {"status": "UP" if self.status else "DOWN"}
+        as_dict |= {child.name: child.to_dict() for child in self.components}
+        as_dict |= {"error": self.error} if self.error else {}
+
+        return as_dict
+
+
+def deep_health_check() -> Dict:
+    """
+    Perform a deep health check of the recommender application. Checks for:
+    - database status (users and services tables have enough members)
+    - celery workers (celery worker is running and responds to scheduled tasks
+    - databus connection (stomp client has a connection to the databus endpoint)
+    - recommender engines (all components can be initialized and loaded)
+    """
+    return HealthStatus(
+        "service",
+        [
+            HealthStatus(
+                "database",
+                [
+                    _check_tables_health(User, HEALTHY_USERS_THRESHOLD),
+                    _check_tables_health(Service, HEALTHY_SERVICES_THRESHOLD),
+                ],
+            ),
+            _check_celery_workers(),
+            _check_databus_connection(),
+            HealthStatus(
+                "recommender_engines",
+                [
+                    _check_engines_health(engine_name, engine_type)
+                    for engine_name, engine_type in ENGINES.items()
+                ],
+            ),
+        ],
+    ).to_dict()
+
+
+def _check_celery_workers() -> HealthStatus:
+    name = "celery_worker"
+    try:
+        ping.delay().get()
+    except Exception:
```

Comment:
I would prefer being precise here -> not using plain Exception

Reply:
ok